File: python/relay.py
import argparse
import logging
import asyncio
import itertools
import os
import socket
import struct
import sys
import time
from Crypto.Util.number import long_to_bytes, bytes_to_long

import config
import dcnet

from cells.null import NullDecoder, NullEncoder
from certify.null import NullAccumulator, NullCertifier
from utils import verbosity
import net.message as m

from elgamal import PublicKey, PrivateKey

logger = logging.getLogger(__file__.rpartition('/')[2])
logger.addHandler(logging.NullHandler())

downmax = 1024 * 64 - 1 # Maximum size of downstream messages
# max amount of data in downstream messages after headers
downcellmax = downmax - m.overhead[m.RELAY_DOWNSTREAM]
# XXX define elsewhere
socks_address = ("localhost", 8080)

""" Remote server interactions """
@asyncio.coroutine
def socks_relay_down(cno, reader, writer, downstream):
    """ Forward data from remote server to downstream queue

    cno (int): connection number
    reader (asyncio.StreamReader): reader for remote server
    writer (asyncio.StreamWriter): writer for remote server (to close if there
      is an error)
    downstream (asyncio.Queue): queue to add downstream data to
    """
    while True:
        yield from asyncio.sleep(0.00000001)
        try:
            buf = yield from reader.read(downcellmax)
        except OSError as e:
            logger.error("socks_relay_down: {}".format(e))
            writer.close()
            return

        logger.debug("socks_relay_down: {} bytes on cno {}"
                    .format(len(buf), cno))
        yield from downstream.put((cno, buf))

        # close the connection to socks relay
        if len(buf) == 0:
            logger.info("socks_relay_down: cno {} closed".format(cno))
            writer.close()
            return

@asyncio.coroutine
def socks_relay_up(cno, reader, writer, upstream):
    """ Send messages from upstream to remote server on cno via writer. """
    while True:
        yield from asyncio.sleep(0.00000001)
        buf = yield from upstream.get()
        dlen = len(buf)

        # client closed connection
        if dlen == 0:
            logger.info("sock_relay_up: closing stream {}".format(cno))
            writer.close()
            return

        logger.debug("socks_relay_up: {} bytes on cno {}".format(dlen, cno))
        try:
            writer.write(buf)
            yield from writer.drain()
        except ConnectionResetError as e:
            logger.error("Could not reach socks proxy: {}".format(e))
            writer.close()
            return
        except OSError as e:
            logger.error("socks_relay_up: {}".format(e))
            writer.close()
            return

""" Other Dissent node interactions """
@asyncio.coroutine
def main_loop(relay, tsocks, crsocks, cwsocks, upstreams, downstream, scheduler):
    """ Handle ciphertext from trustees/clients, forwarding to clients, and
    decoding messages from clients.

    relay (dcnet.Relay): The underlying dcnet Relay
    tsocks (socket list): List of sockets to trustees
    crsocks (socket list): List of connections to read from clients
    cwsocks (socket list): List of connections to write downstream traffic to (a
      combination of elements of crsocks and sockets to access points)
    upstreams (int -> asyncio.Queue()): dict of queues for upstream messages on
      active connections, with connection numbers as keys
    downstream (asyncio.Queue()): queue of messages to send downstream to
      clients
    scheduler (int generator): function to generate the next client index
    """
    loop = asyncio.get_event_loop()
    # branch off two schedulers so trustees can get out ahead
    # XXX tee() can use a lot of memory if one copy gets too far ahead. This
    #   shouldn't be a problem here, but is worth noting
    client_scheduler, trustee_scheduler = itertools.tee(scheduler)

    client_window = 2
    client_inflight = 0

    trustee_window = 10
    trustee_inflight = 0

    up = {'cno':0}
    tmsgs = dict((i, m.pack(m.RELAY_TNEXT, nxt=i)) for i in range(len(crsocks)))

    while True:
        yield from asyncio.sleep(0.00000001)

        # request future cell from trustees
        try:
            nxt = next(trustee_scheduler)
        except StopIteration:
            sys.exit("Scheduler stopped short")

        for tsock in tsocks:
            yield from loop.sock_sendall(tsock, tmsgs[nxt])

        trustee_inflight += 1
        if trustee_inflight < trustee_window:
            continue

        # see if there's anything to send down to clients
        try:
            cno, downbuf = downstream.get_nowait()
        except asyncio.QueueEmpty:
            cno, downbuf = 0, bytearray(0)

        logger.info("downstream to clients: {} bytes on cno {}"
                     .format(len(downbuf), cno)) if (cno > 0) else None

        # send downstream to all clients
        try:
            nxt = next(client_scheduler)
        except StopIteration:
            sys.exit("Scheduler stopped short")

        dbuf = m.pack(m.RELAY_DOWNSTREAM, cno=cno, nxt=nxt, data=downbuf)
        for csock in cwsocks:
            yield from loop.sock_sendall(csock, dbuf)

        client_inflight += 1
        if client_inflight < client_window:
            continue

        # get trustee ciphertexts
        relay.decode_start()
        for tsock in tsocks:
            tslice = yield from loop.sock_recv(tsock, dcnet.cell_length)
            while len(tslice) < dcnet.cell_length:
                tslice += yield from loop.sock_recv(tsock,
                        dcnet.cell_length - len(tslice))
            relay.decode_trustee(tslice)

        # and client upstream ciphertexts
        for csock in crsocks:
            cslice = yield from loop.sock_recv(csock, dcnet.cell_length)
            while len(cslice) < dcnet.cell_length:
                cslice += yield from loop.sock_recv(csock,
                        dcnet.cell_length - len(cslice))
            relay.decode_client(cslice)

        # decode the actual upstream
        outb = relay.decode_cell()
        m.unpack(outb, up)

        client_inflight -= 1
        trustee_inflight -= 1

        # Possibly set up asynchronous sending upstream
        if up['cno'] == 0:
            continue
        else:
            cno = up['cno']
            up['cno'] = 0
        conn = upstreams.get(cno)
        if conn == None:
            # new connection to local socks server
            upstream = asyncio.Queue()
            socks_reader, socks_writer = \
                yield from asyncio.open_connection(*socks_address)
            asyncio.async(socks_relay_down(cno, socks_reader,
                                           socks_writer, downstream))
            asyncio.async(socks_relay_up(cno, socks_reader,
                                         socks_writer, upstream))
            upstreams[cno] = upstream

        yield from upstreams[cno].put(up['data'])


""" Setup """
def main():
    logging.basicConfig()
    p = argparse.ArgumentParser(description="Basic DC-net relay")
    p.add_argument("-p", "--port", type=int, help="Port to listen for \
                   connections on", required=True, dest="port")
    p.add_argument("config_dir")
    p.add_argument("-s", "--socks", type=str, metavar="host:port",
                   help="SOCKS proxy address",
                   default="localhost:8080", dest="socks_addr")
    p.add_argument("-v", type=str, help="display more output (default: WARN)",
                   choices=verbosity.keys(), default="WARN", dest="verbose")
    opts = p.parse_args()
    logger.setLevel(verbosity[opts.verbose])

    global socks_address
    saddr, sport = opts.socks_addr.split(":")
    socks_address = saddr, int(sport)

    system_config = config.load(config.SystemConfig,
                                os.path.join(opts.config_dir, "system.json"))
    nclients = len(system_config.clients.ids)
    ntrustees = len(system_config.trustees.ids)
    naps = len(system_config.aps.ids)

    # start up a new relay
    relay = dcnet.Relay(ntrustees, NullAccumulator(), NullDecoder())
    relay.add_nyms(nclients)
    relay.sync(None)

    # server socket
    print("Starting relay on {}".format(opts.port))
    ssock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    ssock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    ssock.bind(("0.0.0.0", opts.port))
    ssock.listen(1024)

    # make sure everybody connects
    print(("Waiting for {} clients, {} trustees, and {} access points")
          .format(nclients, ntrustees, naps))
    ccli, ctru, caps = 0, 0, 0
    crsocks = [None] * nclients
    cwsocks = []
    tsocks = [None] * ntrustees
    apsocks = [None] * naps
    while ccli < nclients or ctru < ntrustees or caps < naps:
        conn, addr = ssock.accept()
        buf = bytearray(1)
        buf[0] = kind = bytes_to_long(conn.recv(1))
        buf.extend(conn.recv(m.sizes[kind] - 1))
        reg = {}
        m.unpack(buf, reg)
        conn.setblocking(0)

        if reg['kind'] == m.TRUSTEE_CONNECT and ctru < ntrustees:
            if tsocks[reg['node']] is not None:
                sys.exit("Trustee connected twice")
            tsocks[reg['node']] = conn
            ctru += 1
        elif reg['kind'] == m.CLIENT_CONNECT and ccli < nclients:
            if crsocks[reg['node']] is not None:
                sys.exit("Clients connected twice")
            crsocks[reg['node']] = conn
            if reg['ap'] == -1:
                cwsocks.append(conn)
            ccli += 1
        elif reg['kind'] == m.AP_CONNECT and caps < naps:
            if apsocks[reg['node']] is not None:
                sys.exit("Access point connected twice")
            apsocks[reg['node']] = conn
            cwsocks.append(conn)
            caps += 1
            global downcellmax
            # Leave room for IP/UDP headers and AP headers
            downcellmax = downmax - 28 - m.overhead[m.AP_DOWNSTREAM] - \
                                        m.overhead[m.RELAY_DOWNSTREAM]
        else:
            sys.exit("Illegal node number or connection type")
    print("All clients, trustees, and access points connected")

    upstreams = {}
    downstream = asyncio.Queue()
    scheduler = itertools.cycle(range(nclients))

    # start the main relay loop
    asyncio.async(main_loop(relay, tsocks, crsocks, cwsocks,
                            upstreams, downstream, scheduler))
    loop = asyncio.get_event_loop()
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        pass
    loop.close()


if __name__ == "__main__":
    main()

File: dissent/relay.go
package main

import (
	"io"
	"fmt"
	"net"
	"log"
	"time"
	"encoding/binary"
	"github.com/dedis/crypto/abstract"
	"dissent/dcnet"
)

type Trustee struct {
	pubkey abstract.Point
}

type AnonSet struct {
	suite abstract.Suite
	trustees []Trustee
}

func startRelay() {
	tg := dcnet.TestSetup(suite, factory, nclients, ntrustees)
	me := tg.Relay

	// Start our own local HTTP proxy for simplicity.
/*
	go func() {
		proxy := goproxy.NewProxyHttpServer()
		proxy.Verbose = true
		println("Starting HTTP proxy")
		log.Fatal(http.ListenAndServe(":8888", proxy))
	}()
*/

	lsock,err := net.Listen("tcp", bindport)
	if err != nil {
		panic("Can't open listen socket:"+err.Error())
	}

	// Wait for all the clients and trustees to connect
	ccli := 0
	ctru := 0
	csock := make([]net.Conn, nclients)
	tsock := make([]net.Conn, ntrustees)
	for ; ccli < nclients || ctru < ntrustees ; {
		fmt.Printf("Waiting for %d clients, %d trustees\n",
				nclients-ccli, ntrustees-ctru)

		conn,err := lsock.Accept()
		if err != nil {
			panic("Listen error:"+err.Error())
		}

		b := make([]byte,1)
		n,err := conn.Read(b)
		if n < 1 || err != nil {
			panic("Read error:"+err.Error())
		}

		node := int(b[0] & 0x7f)
		if b[0] & 0x80 == 0 && node < nclients {
			if csock[node] != nil {
				panic("Oops, client connected twice")
			}
			csock[node] = conn
			ccli++
		} else if b[0] & 0x80 != 0 && node < ntrustees {
			if tsock[node] != nil {
				panic("Oops, trustee connected twice")
			}
			tsock[node] = conn
			ctru++
		} else {
			panic("illegal node number")
		}
	}
	println("All clients and trustees connected")

	// Create ciphertext slice buffers for all clients and trustees
	clisize := me.Coder.ClientCellSize(payloadlen)
	cslice := make([][]byte, nclients)
	for i := 0; i < nclients; i++ {
		cslice[i] = make([]byte, clisize)
	}
	trusize := me.Coder.TrusteeCellSize(payloadlen)
	tslice := make([][]byte, ntrustees)
	for i := 0; i < ntrustees; i++ {
		tslice[i] = make([]byte, trusize)
	}

	// Periodic stats reporting
	begin := time.Now()
	report := begin
	period,_ := time.ParseDuration("3s")
	totupcells := int64(0)
	totupbytes := int64(0)
	totdowncells := int64(0)
	totdownbytes := int64(0)

	conns := make(map[int] chan<- []byte)
	downstream := make(chan connbuf)
	nulldown := connbuf{}	// default empty downstream cell
	window := 2		// Maximum cells in-flight
	inflight := 0		// Current cells in-flight
	for {
		//print(".")

		// Show periodic reports
		now := time.Now()
		if now.After(report) {
			duration := now.Sub(begin).Seconds()
			fmt.Printf("@ %f sec: %d cells, %f cells/sec, %d upbytes, %f upbytes/sec, %d downbytes, %f downbytes/sec\n",
				duration,
				totupcells, float64(totupcells) / duration,
				totupbytes, float64(totupbytes) / duration,
				totdownbytes, float64(totdownbytes) / duration)

			// Next report time
			report = now.Add(period)
		}

		// See if there's any downstream data to forward.
		var downbuf connbuf
		select {
		case downbuf = <-downstream: // some data to forward downstream
			//fmt.Printf("v %d\n", len(dbuf)-6)
		default:		// nothing at the moment to forward
			downbuf = nulldown
		}
		dlen := len(downbuf.buf)
		dbuf := make([]byte, 6+dlen)
		binary.BigEndian.PutUint32(dbuf[0:4], uint32(downbuf.cno))
		binary.BigEndian.PutUint16(dbuf[4:6], uint16(dlen))
		copy(dbuf[6:], downbuf.buf)

		// Broadcast the downstream data to all clients.
		for i := 0; i < nclients; i++ {
			//fmt.Printf("client %d -> %d downstream bytes\n",
			//		i, len(dbuf)-6)
			n,err := csock[i].Write(dbuf)
			if n != 6+dlen {
				panic("Write to client: "+err.Error())
			}
		}
		totdowncells++
		totdownbytes += int64(dlen)
		//fmt.Printf("sent %d downstream cells, %d bytes \n",
		//		totdowncells, totdownbytes)

		inflight++
		if inflight < window {
			continue	// Get more cells in flight
		}

		me.Coder.DecodeStart(payloadlen, me.Histoream)

		// Collect a cell ciphertext from each trustee
		for i := 0; i < ntrustees; i++ {
			n,err := io.ReadFull(tsock[i], tslice[i])
			if n < trusize {
				panic("Read from client: "+err.Error())
			}
			//println("trustee slice")
			//println(hex.Dump(tslice[i]))
			me.Coder.DecodeTrustee(tslice[i])
		}

		// Collect an upstream ciphertext from each client
		for i := 0; i < nclients; i++ {
			n,err := io.ReadFull(csock[i], cslice[i])
			if n < clisize {
				panic("Read from client: "+err.Error())
			}
			//println("client slice")
			//println(hex.Dump(cslice[i]))
			me.Coder.DecodeClient(cslice[i])
		}

		outb := me.Coder.DecodeCell()
		inflight--

		totupcells++
		totupbytes += int64(payloadlen)
		//fmt.Printf("received %d upstream cells, %d bytes\n",
		//		totupcells, totupbytes)

		// Process the decoded cell
		if outb == nil {
			continue	// empty or corrupt upstream cell
		}
		if len(outb) != payloadlen {
			panic("DecodeCell produced wrong-size payload")
		}

		// Decode the upstream cell header (may be empty, all zeros)
		cno := int(binary.BigEndian.Uint32(outb[0:4]))
		uplen := int(binary.BigEndian.Uint16(outb[4:6]))
		//fmt.Printf("^ %d (conn %d)\n", uplen, cno)
		if cno == 0 {
			continue	// no upstream data
		}
		conn := conns[cno]
		if conn == nil {	// client initiating new connection
			conn = relayNewConn(cno, downstream)
			conns[cno] = conn
		}
		if 6+uplen > payloadlen {
			log.Printf("upstream cell invalid length %d", 6+uplen)
			continue
		}
		conn <- outb[6:6+uplen]
	}
}

